### PR TITLE
Refactor url handling in web.py

### PR DESCRIPTION
**Pull request by fstachura**
This PR aims to refactor how Elixir routes urls in web.py. The main focus were lines [55-123](https://github.com/bootlin/elixir/blob/master/http/web.py#L55-L123), although I also decided to move the code that handles different "routes" to new functions, and this caused some more changes to be necessary. 

Some notes:
* The main idea behind the patch is that different code paths should handle different routes (ident and source in this case) and that the code that handles different routes shouldn't directly write the HTTP response, just prepare information necessary to create the response. Routing starts in `route`, and the function returns just the status code and (if applicable) contents of the response. Only cases currently used by Elixir are handled. 
* [Filters](https://github.com/bootlin/elixir/tree/master/http/filters) are scripts that transform source to, for example, create clickable links for identifiers. Filters access global scope of web.py to gather information and store data. This is not great, because it's hard to know what of information filters require to work, and we can never be sure if filters won't modify something in the global scope of web.py. This fact caused problems when moving code that generates source response to a function - many global variables suddenly became local. The solution to this was to explicitly create a new, separate global context for filters and pass it to [exec](https://docs.python.org/3/library/functions.html#exec). Now context for filters is defined in a single place and filters cannot mess with global scope of web.py. Current version of filters seems to somewhat approximate classes (functions associated with "encapsulated" data), but redesigning them is out of scope. 
* I deliberately left as much of the code that generates HTML untouched. This PR mainly focuses only on routing logic and response handling.
* I added the latest tag redirect requested in #261. This perhaps should've been a different PR, but I started refactoring when I was trying to add this feature. 
* It may seem that I'm handling errors with return values in, for example, route. This is because I don't think using exceptions for anything else than errors is a good idea, and IMO a redirect is not an error. I also thought about designing this in a kind of functional way, where a list of functions was executed one after another, unless one of the functions returned a valid response code. I abandoned this idea because it seemed way too over engineered and had some issues related to building state. And if we move to WSGI, most of this will be probably be handled by a framework. Hence, I would not spend more time trying to perfectly design this part. 
* L158-L182 was moved to generate_versions, L128-L132 to get_projects, both almost without any changes

### CODE REVIEW

File: http/web.py [hunk 1]
```diff
@@ -18,6 +18,11 @@
 #  You should have received a copy of the GNU Affero General Public License
 #  along with Elixir.  If not, see <http://www.gnu.org/licenses/>.
 
+# prepare a default globals dict to be later used for filter context
+default_globals = {
+    **globals(),
+}
```
Comment (tleb) [inline http/web.py:24]: I like the idea, but it can be added when it becomes useful and only at that point.

File: http/web.py [hunk 2]
```diff
@@ -34,9 +39,14 @@ def print(arg, end='\n'):
 import re
 from re import search, sub
 
+import jinja2
+loader = jinja2.FileSystemLoader(os.path.join(os.path.dirname(os.path.realpath(__file__)), '../templates/'))
```
Comment (tleb) [inline http/web.py:43]: Long line, please split the path into `templates_dir_path` or something of the sort.

Also, let's try to stick with (1) file header comment, (2) includes, (3) code. So please group imports together.

File: http/web.py [hunk 3]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
 # Enable CGI Trackback Manager for debugging (https://docs.python.org/fr/3/library/cgitb.html)
 cgitb.enable(display=0, logdir=errdir, format='text')
 
+form = cgi.FieldStorage()
```
Comment (tleb) [inline http/web.py:60]: I'd prefer if the name was not generic and was not being shadowed by local variables inside functions.

> Use descriptive names for globals, short names for locals
> *— Rob Pike*

[Related](https://research.swtch.com/names).

File: http/web.py [hunk 4]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
-    arg = m.group(5)
-
-    if not validFamily(family):
-        family = 'C'
-
-    search_family = 'A'
-
-    basedir = os.environ['LXR_PROJ_DIR']
+# get query class from basedir and project is paths exist
+def get_query(basedir, project):
```
Comment (tleb) [inline http/web.py:66]: Please be more explicit in your comments. Something like:

```python
# Return a Query class instance (ie an object representing a database)
# basedir: absolute path to parent directory of all repos and databases
# project: project name, matching projects/*.sh (eg "linux")
```

File: http/web.py [hunk 5]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
-    break
-projects.sort()
+        return None
 
-from query import Query
+    return Query(datadir, repodir)
+
+
+# parse source path
+def parse_source_path(path):
```
Comment (tleb) [inline http/web.py:77]: Comment here is exactly the function name, so not useful. I guess `path` here means the HTTP path but it is ambiguous because we deal with HTTP paths and filesystem paths in this `http/web.py` file.

About the return value: you have [`collections.namedtuple`](https://docs.python.org/3/library/collections.html) which allows a nicer interaction:

```python
ParsedHTTPPath = collections.namedtuple('ParsedHTTPPath', 'project, version, arg')
p = ParsedHTTPPath(project='a', version='b', arg='c')

# then use with p. syntax, no p['']
assert p.project == 'c'
```

File: http/web.py [hunk 6]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
-    'version': version,
-    'url': url,
-    'project': project,
-    'projects': projects,
-    'ident': ident,
-    'family': search_family,
-    'breadcrumb': '<a class="project" href="'+version+'/source">/</a>'
-}
+# return 301 to actual latest version if version in parsed source url is latest
+def redirect_source_on_latest(parsed_path, q):
```
Comment (tleb) [inline http/web.py:95]: This feature typically shouldn't be too hard to move into a separate commit.

File: http/web.py [hunk 7]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+    if parsed_path['version'] == 'latest':
+        parsed_path['version'] = parse.quote(q.query('latest'))
```
Comment (tleb) [inline http/web.py:97]: Are you not double quoting here? `stringify_source_path` does the same `parse.quote(version)` call.

File: http/web.py [hunk 8]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
-            for _tag in tags:
-                _tag_encoded = parse.quote(_tag, safe='')
-                if _tag == tag: v += '\t\t\t\t<li class="li-link active"><a href="'+_tag_encoded+'/'+url+'">'+_tag+'</a></li>\n'
-                else: v += '\t\t\t\t<li class="li-link"><a href="'+_tag_encoded+'/'+url+'">'+_tag+'</a></li>\n'
-            v += '\t\t\t</ul></li>\n'
-    v += '\t</ul></li>\n'
+# return 301 if path contains a trailing slash
+def redirect_on_trailing_slash(path):
+    if path[-1] == '/':
+        return (301, path.rstrip('/'))
```
Comment (tleb) [inline http/web.py:103]: Side note: I like the shape of functions that return a tuple or `None`. Code is straight forward.

About `redirect_on_trailing_slash()`: any reason to redirect rather to handle as if no slash was present? The whole [robustness principle](https://en.wikipedia.org/wiki/Robustness_principle).

File: http/web.py [hunk 9]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+
+# handle source url
+def handle_source_url(path, form):
+    status = redirect_on_trailing_slash(path)
+    if status is not None:
+        return status
+
+    parsed_path = parse_source_path(path)
+    if parsed_path is None:
+        return (400,)
```
Comment (tleb) [inline http/web.py:113]: Is this not a 404? If we fail parsing it means it doesn't have the shape we expect, so it is not a HTTP path we know about.

File: http/web.py [hunk 10]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+        parsed_path = {
+            'project': m.group(1),
+            'version': m.group(2),
+            'arg': m.group(4),
+        }
+
+        return parsed_path
+
+# turn parsed source path to string
+def stringify_source_path(ppath):
```
[the remark this answers is not in the document]
Comment (tleb) [reply http/web.py:90]: nit: why capture group 3 even though it is not used?

File: http/web.py [hunk 11]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+    parsed_path = parse_source_path(path)
+    if parsed_path is None:
+        return (400,)
+
+    query = get_query(os.environ['LXR_PROJ_DIR'], parsed_path['project'])
+    if not query:
+        return (400,)
+
+    if not search('^[A-Za-z0-9_/.,+-]*$', parsed_path['arg']):
+        return (400,)
```
Comment (tleb) [inline http/web.py:120]: This check is done here in `handle_source_url()` but not in the other function that calls `get_query()` which is `handle_ident_url()`. If it is normal, a small comment could be useful. If not, it could be moved into `get_query()`.

File: http/web.py [hunk 12]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+    status = redirect_source_on_latest(parsed_path, query)
+    if status is not None:
+        return status
+
+    url = 'source' + parsed_path['arg']
+    return generate_source_page(query, url, os.environ['LXR_PROJ_DIR'], parsed_path)
+
+
+# parse ident path
+def parse_ident_path(path):
```
Comment (tleb) [inline http/web.py:131]: Again, empty comment.

File: http/web.py [hunk 13]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+    if m:
+        parsed_path = {
+            'project': m.group(1),
+            'version': m.group(2),
+            'family': str(m.group(3)).upper(),
+            'arg': m.group(5),
+        }
+
+        if not validFamily(parsed_path['family']):
+            parsed_path['family'] = 'C'
```
Comment (tleb) [inline http/web.py:143]: Is this useful in practice for some families or it is just to be lean in what we accept? A comment would be nice.

File: http/web.py [hunk 14]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+
+# handle ident search post request by redirecting to ident/$ident_name
+def handle_ident_post_form(parsed_path, form):
+    post_ident = form.getvalue('i')
+    post_family = str(form.getvalue('f')).upper()
+
+    if not validFamily(post_family):
+        post_family = 'C'
+
+    if parsed_path.get('ident', '') == '' and post_ident:
```
Comment (tleb) [inline http/web.py:160]: Is it possible for `'ident' in parsed_path` to be true? It doesn't look like it is.

File: http/web.py [hunk 15]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+    path = f'/{ppath["project"]}/{parse.quote(ppath["version"])}/{ppath["family"]}/ident/{ppath["arg"]}'
+    return path.rstrip('/')
+
+# handle ident search post request by redirecting to ident/$ident_name
+def handle_ident_post_form(parsed_path, form):
+    post_ident = form.getvalue('i')
+    post_family = str(form.getvalue('f')).upper()
+
+    if not validFamily(post_family):
+        post_family = 'C'
```
Comment (tleb) [inline http/web.py:158]: Same question: is this useful? If yes, please add a comment to explain the context.

File: http/web.py [hunk 16]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+        post_ident = parse.quote(post_ident.strip(), safe='/')
+        parsed_path['family'] = post_family
+        parsed_path['arg'] = post_ident
+        return (302, stringify_ident_path(parsed_path))
+
+# return 301 to actual latest version if version in parsed ident url is latest
+def redirect_ident_on_latest(parsed_path, q):
+    if parsed_path['version'] == 'latest':
+        parsed_path['version'] = parse.quote(q.query('latest'))
+        return (301, stringify_ident_path(parsed_path))
```
Comment (tleb) [inline http/web.py:170]: Is this not the same thing as the `redirect_source_on_latest()` function? To be honest both functions could be inlined into their parent functions.

File: http/web.py [hunk 17]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+                v += '\t\t\t<span>'+submenu+'</span>\n'
+                v += '\t\t\t<ul>\n'
+                for _tag in tags:
+                    _tag_encoded = parse.quote(_tag, safe='')
+                    if _tag == tag: v += '\t\t\t\t<li class="li-link active"><a href="'+_tag_encoded+'/'+url+'">'+_tag+'</a></li>\n'
+                    else: v += '\t\t\t\t<li class="li-link"><a href="'+_tag_encoded+'/'+url+'">'+_tag+'</a></li>\n'
+                v += '\t\t\t</ul></li>\n'
+        v += '\t</ul></li>\n'
+
+    return v
```
Comment (tleb) [inline http/web.py:241]: This code is error prone. Can we build a datastructure that we pass to the template? We have a full jinja2 library included and we build HTML by hand.

File: http/web.py [hunk 18]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+    data = {
+        'baseurl': '/' + project + '/',
+        'tag': tag,
+        'version': version,
+        'url': url,
+        'project': project,
+        'projects': projects,
+        'ident': ident,
+        'family': search_family,
+        'breadcrumb': '<a class="project" href="'+version+'/source">/</a>'
```
Comment (tleb) [inline http/web.py:264]: HTML built by hand.

File: http/web.py [hunk 19]
```diff
@@ -274,16 +354,35 @@ def print(arg, end='\n'):
...
+            "family": family,
+            "path": path,
+            "tag": tag,
+            "q": q,
+        }
+
         # Source common filter definitions
         os.chdir('filters')
-        exec(open("common.py").read())
+        exec(open("common.py").read(), filter_ctx)
```
Comment (tleb) [inline http/web.py:376]: Ah so we can pass a global context to `exec()` calls. It is a bit better than what we had previously. Why did you not go with your suggestion of removing those calls and going with standard imports?

File: http/web.py [hunk 20]
```diff
@@ -329,8 +429,37 @@ def print(arg, end='\n'):
...
+        'url': url,
+        'project': project,
+        'projects': projects,
+        'ident': ident,
+        'family': search_family,
+        'breadcrumb': '<a class="project" href="'+version+'/source">/</a>'
+    }
+
     data['title'] = ident+' identifier - '+title_suffix
+    data['versions'] = generate_versions(q.query('versions'), url, tag)
```
Comment (tleb) [inline http/web.py:462]: `data` is built just above, you could add the fields in the dict declaration.

File: http/web.py (hunk 17 above)
Comment (fstachura) [reply http/web.py:241]: We could, but this is left as close as possible to the original. I definitely want to move all html building to templates. Let's refactor this in a separate PR.

File: http/web.py [hunk 21]
```diff
@@ -444,17 +573,31 @@ def print(arg, end='\n'):
...
+        realprint('Status: 302 Found')
+        realprint('Location: '+ result[1] +'\n')
+        exit()
+    elif result[0] == 400:
+        realprint('Status: 400 Bad Request\n')
+        exit()
+    elif result[0] == 404:
+        realprint('Status: 404 Not Found')
+        realprint('Content-Type: text/html;charset=utf-8\n')
+        realprint(result[1], end='')
```
Comment (tleb) [inline http/web.py:602]: It would be nice to handle all non `2xx` HTTP codes by a 500. Else if we forget one case here we return nothing.

File: http/web.py (hunk 19 above)
Comment (fstachura) [reply http/web.py:376]: I don't want this PR to grow too big. And I doubt that just changing exec to import will work - filters share some state between each other. Properly retesting filters once we refactor them might take some effort.

File: http/web.py (hunk 18 above)
Comment (fstachura) [reply http/web.py:264]: Again, would rather address this in a separate PR, where I move all HTML building to templates. This is left as in the original.

File: http/web.py (hunk 19 above)
Comment (tleb) [reply http/web.py:376]: Yes I reread your PR message and understand the explanation. Makes sense.

File: http/web.py (hunk 16 above)
Comment (fstachura) [reply http/web.py:170]: Pretty much the same thing, except uses a different path stringifier. I will move it to the caller function.

**Comment by tleb** [comment]
That's good stuff. Agreed templating rework can wait, or it can be this PR but in separate commits.

I like the way functions return a tuple. Agreed, functions pipelines would be overengineering.

File: http/web.py (hunk 9 above)
Comment (fstachura) [reply http/web.py:113]: Left as in the original. https://github.com/bootlin/elixir/blob/master/http/web.py#L109
And in this case it means it somehow failed parsing despite matching in route. 
But I guess 404 does make a bit more sense still.

File: http/web.py (hunk 13 above)
Comment (fstachura) [reply http/web.py:143]: This mostly backward compatibility - the original code replaces ident family with a default if family is not recognized. https://github.com/bootlin/elixir/blob/master/http/web.py#L66
I won't argue whether this makes sense or not as I didn't read the ident response generation code too much yet. I would leave this as it is for now - my goal was to refactor while leaving as much of the original behavior as possible, regardless of whether it seems to make sense or not.

File: http/web.py (hunk 21 above)
Comment (fstachura) [reply http/web.py:602]: I assume you mean all codes unknown to the handler?

File: http/web.py (hunk 1 above)
Comment (fstachura) [reply http/web.py:24]: As noted later, this is used to better separate global context of filter scripts.

File: http/web.py (hunk 6 above)
Comment (fstachura) [reply http/web.py:95]: Maybe, but I would have to replace it with this first: https://github.com/bootlin/elixir/blob/master/http/web.py#L139
Just to replace it with the new logic again. Let's maybe leave it as it is.

File: http/web.py (hunk 7 above)
Comment (fstachura) [reply http/web.py:94]: Good catch

File: http/web.py (hunk 5 above)
Comment (fstachura) [reply http/web.py]: Namedtuples are a good idea, completely forgot about them

File: http/web.py (hunk 21 above)
Comment (fstachura) [reply http/web.py:615]: @tleb I implemented handling for route returning None and unknown error codes

File: http/web.py (hunk 20 above)
Comment (fstachura) [reply http/web.py]: Moved data assignments to dict

File: http/web.py (hunk 11 above)
Comment (fstachura) [reply http/web.py:116]: This is just argument validation, I added a comment. Not sure what get_query has to do with this - it doesn't use arg (path/ident).

File: http/web.py (hunk 9 above)
Comment (fstachura) [reply http/web.py]: Revised status codes and added plain text error messages. Once HTML generation is refactored, plain text will be replaced by HTML. Old web.py returned just empty responses which were replaced by default error messages in most browsers.

File: http/web.py (hunk 15 above)
Comment (fstachura) [reply http/web.py]: Since `family` is replaced with a default in `parse_ident_path`, I'm pretty sure this cannot be triggered just from the site. You would have to use curl or modify a request from the network tab. So I don't know if it's worth leaving in. If I'm right, removing this could only potentially break a poorly coded scrapper.

File: http/web.py [hunk 22]
```diff
@@ -47,145 +57,215 @@ def print(arg, end='\n'):
...
+        return handle_ident_url(path, form)
+    else:
+        return (400,)
+
+
+def get_projects(basedir):
+    projects = []
+    for (dirpath, dirnames, filenames) in os.walk(basedir):
+        projects.extend(dirnames)
+        break
```
[the remark this answers is not in the document]
Comment (fstachura) [reply http/web.py:220]: > It means we hope that os.walk always returns the root directory first.

But shouldn't it always do that? `topdown` is `True` by default https://docs.python.org/3.8/library/os.html#os.walk

inb4 I agree that the current code is kind of awkward. Your version throws if basedir does not exist which I like a bit more than returning an empty list (what happens now).

File: http/web.py [hunk 23]
```diff
@@ -148,7 +148,7 @@ def parse_ident_path(path):
             m.group(1),
             m.group(2),
             family,
-            m.group(4)
+            m.group(4)[1:]
```
Comment (tleb) [inline http/web.py:151]: I think this can throw an exception as the last group is `(.*)`.

**Comment by tleb** [comment]
Thanks for the nice commit log!

File: http/web.py (hunk 23 above)
Comment (fstachura) [reply http/web.py:151]: Empty groups seem to be an empty string, not a None.

```
>>> m = search('^/([^/]*)/([^/]*)(?:/([^/]))?/([^/]*)(.*)$', "/a/b/c/d")
>>> m.group(5)
''
```

Comment (tleb) [reply http/web.py:151]: Perfect, thanks!

**Comment by tleb** [comment]
After some testing, applied on [production](https://github.com/bootlin/elixir/tree/prod). Thanks Franek!

**Comment by miquelraynal** [comment]
Excellent! No more dead links \o/
